README.md
<p align="center">
  <img src="ColabKit.jpg" alt="ColabKit logo" width="250" height="250">
</p>

---

# ColabKit

We have officially renamed the PyColab library to ColabKit to avoid conflicts with another existing package on PyPI. We appreciate your understanding and continued support.

ColabKit is a Python library designed to enhance the experience of working in Jupyter Notebook and Google Colab environments. With ColabKit, you can simplify common tasks, manipulate media, record audio, and create interactive widgets more efficiently.

## Installation

You can easily install ColabKit using `pip`. Open a terminal and run:

```bash
pip install ColabKit
```

## Key Features

- **Media Manipulation**: Visualize and edit video and audio files directly in your notebook.
- **Audio Recording**: Record real-time audio through your browser with a simple function.
- **Interactive Widgets**: Create interactive buttons in the notebook to execute custom code.
- **System Utilities**: Simplify common tasks, such as file deletion and file existence checks.

## Usage Examples

### Video Display

```python
from ColabKit.media import showMedia

# Display a video file
showMedia('video.mp4', 'video')
```

### Audio Recording

```python
from ColabKit.record import record

# Record audio and save the result to a WAV file
record('recording.wav')
```

### Creating Interactive Buttons

```python
from ColabKit.widget import generate_button

# Create an interactive button
def my_function():
    print("You clicked the button!")

generate_button("Click Me", "Click to execute my_function", "success", my_function)
```

## Documentation

For more information and detailed documentation, visit [the ColabKit repository on GitHub](https://github.com/wipodev/ColabKit).

## Contributions

We appreciate contributions! If you wish to contribute or report issues, visit [our GitHub repository](https://github.com/wipodev/ColabKit).

## License

This project is under the MIT License. See the [LICENSE](https://github.com/wipodev/ColabKit/blob/main/LICENSE) file for more details.

---


CHANGELOG.md
# Lanzamiento con Cambio de Nombre

En este lanzamiento, ColabKit (anteriormente PyColab) te trae las últimas actualizaciones y características en una librería de Python diseñada para simplificar tu trabajo en entornos de Google Colab. ColabKit ofrece una colección de módulos que incluyen herramientas para la manipulación de medios, la grabación de audio, la gestión de archivos y la creación de widgets interactivos en tus cuadernos de Colab.

**Aspectos Destacados:**

- `media`: Visualiza y procesa medios en tus cuadernos de Colab.
- `record`: Graba y administra grabaciones de audio directamente desde tu navegador.
- `utils`: Utilidades para la gestión de archivos y la interacción con Google Colab.
- `widget`: Crea botones y widgets interactivos en tus cuadernos.

**Cambio de Nombre:**

En esta versión, hemos rebautizado oficialmente la librería de PyColab a ColabKit para evitar conflictos con otro paquete existente en PyPI. Agradecemos tu comprensión y apoyo continuo.

Estamos emocionados de compartir estas actualizaciones con la comunidad de Python y Colab. ColabKit tiene como objetivo ayudarte en tus proyectos y experimentos en Google Colab. No dudes en proporcionar comentarios, informar problemas o contribuir al desarrollo de la librería en nuestro repositorio de GitHub.

**Documentación y Uso:**

Para obtener más detalles sobre cómo utilizar ColabKit y acceder a la documentación completa, visita el [enlace a la documentación](https://github.com/wipodev/ColabKit/tree/main/docs).

**Notas Adicionales:**

- Mantén tu instalación de ColabKit actualizada utilizando `pip install colabkit --upgrade`.
- Si tienes alguna pregunta o encuentras problemas, no dudes en abrir un "Issue" en nuestro repositorio de GitHub para recibir soporte.

Gracias por tu continuo apoyo y entusiasmo por ColabKit. ¡Esperamos que encuentres esta librería valiosa en tus proyectos!


tests/test_media.py
import unittest
from unittest.mock import patch, MagicMock
from src.ColabKit import media
import ffmpeg

class MediaTestCase(unittest.TestCase):


    @patch('src.ColabKit.media.VideoFileClip')
    def test_video_duration(self, mock_video_file_clip):
        # Test case for a video file with a duration of 10.5 seconds
        mock_instance = mock_video_file_clip.return_value
        mock_instance.duration = 10.5
        duration = media.video_duration('video_path.mp4')
        # Assert that the duration is correct
        self.assertEqual(duration, 10.5)

    @patch('src.ColabKit.media.VideoFileClip')
    def test_get_video_resolution(self, mock_video_file_clip):
        # Test case for a video file with a resolution of 1920x1080
        mock_instance = mock_video_file_clip.return_value
        mock_instance.size = (1920, 1080)
        resolution = media.get_video_resolution('video_path.mp4')
        # Assert that the resolution is correct
        self.assertEqual(resolution, (1920, 1080))

    def test_ffmpeg_err(self):
        # Test case for an error message from FFmpeg
        stderr_output = b"Error: Something went wrong\nAnother comment occurred"
        formatted_output = media.ffmpeg_err(stderr_output)
        expected_output = "\033[43m\033[101mError: Something went wrong\033[0m"
        # Assert that the error message is correct after formatting
        self.assertEqual(formatted_output, expected_output)

    @patch('ffmpeg.run')
    @patch('builtins.print')
    @patch('src.ColabKit.media.remove_file')
    def test_ffmpeg_proc_success(self, mock_remove_file, mock_print, mock_run):
        # Test case for a successful FFmpeg process
        mock_run.return_value = None
        result = media.ffmpeg_proc('output_stream', 'output_file', 'message')
        # Assert that the function returned True after executing the FFmpeg process
        self.assertTrue(result)
        # Assert that the print function was called with the expected message
        mock_print.assert_called_with('File message and saved as output_file')
        # Assert that the remove_file function was not called
        mock_remove_file.assert_not_called()

    @patch('ffmpeg.run', side_effect=ffmpeg.Error("", "", stderr=b'Error occurred'))
    @patch('builtins.print')
    @patch('src.ColabKit.media.remove_file')
    def test_ffmpeg_proc_failure(self, mock_remove_file, mock_print, mock_run):
        # Test case for a failed FFmpeg process 
        result = media.ffmpeg_proc('output_stream', 'output_file', 'message')
        # Assert that the function returned False after executing the FFmpeg process
        self.assertFalse(result)
        # Assert that the print function was called with the expected message
        mock_print.assert_called_with(media.ffmpeg_err(b'Error occurred'))
        # Assert that the remove_file function was called
        mock_remove_file.assert_called_with('output_file')

    @patch('src.ColabKit.media.ffmpeg_proc', return_value=True)
    @patch('builtins.print')
    def test_ffmpeg_conv_success(self, mock_print, mock_ffmpeg_proc):
        # Test case for a successful conversion
        result = media.ffmpeg_conv('input_file.mp4', 'output_file.mp4')
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("Converting input_file.mp4 to output_file.mp4...")
        # Assert that the remove_file function was called
        self.assertTrue(result)

    @patch('src.ColabKit.media.ffmpeg_proc', return_value=False)
    @patch('builtins.print')
    def test_ffmpeg_conv_failure(self, mock_print, mock_ffmpeg_proc):
        # Test case for a failed conversion 
        result = media.ffmpeg_conv('input_file.mp4', 'output_file.mp4')
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with("Converting input_file.mp4 to output_file.mp4...")
        # Assert that the remove_file function was called
        self.assertFalse(result)

    @patch('src.ColabKit.media.ffmpeg_proc', return_value=True)
    @patch('builtins.print')
    def test_ffmpeg_trim_success(self, mock_print, mock_ffmpeg_proc):
        # Test case for a successful trim
        result = media.ffmpeg_trim('input_file.mp4', 10, 5, 'output_file.mp4')
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("Trimming input_file.mp4 to 5 seconds...")
        # Assert that the remove_file function was called
        self.assertTrue(result)

    @patch('src.ColabKit.media.ffmpeg_proc', return_value=False)
    @patch('builtins.print')
    def test_ffmpeg_trim_failure(self, mock_print, mock_ffmpeg_proc):
        # Test case for a failed trim
        result = media.ffmpeg_trim('input_file.mp4', 10, 5, 'output_file.mp4')
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with("Trimming input_file.mp4 to 5 seconds...")
        # Assert that the remove_file function was called
        self.assertFalse(result)

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1920, 1080))
    @patch('src.ColabKit.media.ffmpeg_proc', return_value=True)
    @patch('builtins.print')
    def test_resize_video_success(self, mock_print, mock_ffmpeg_proc, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("Resizing video to 720p...")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video_720p.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1280, 720))
    @patch('src.ColabKit.media.ffmpeg_proc', return_value=True)
    @patch('builtins.print')
    def test_resize_video_small_resolution(self, mock_print, mock_ffmpeg_proc, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("No need to change the size")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(640, 360))
    @patch('builtins.print')
    def test_resize_video_no_need_to_change(self, mock_print, mock_get_video_resolution):
        # Test case for a successful resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned True after executing the FFmpeg process
        mock_print.assert_called_with("No need to change the size")
        # Assert that the remove_file function was called
        self.assertEqual(result, 'input_video.mp4')
        # Assert that the get_video_resolution function was called
        mock_get_video_resolution.assert_called_once_with('input_video.mp4')

    @patch('src.ColabKit.media.get_video_resolution', return_value=(1920, 1080))
    @patch('src.ColabKit.media.ffmpeg_proc', return_value=False)
    @patch('builtins.print')
    def test_resize_video_failure(self, mock_print, mock_ffmpeg_proc, mock_get_video_resolution):
        # Test case for a failed resize
        result = media.resize_video('input_video.mp4', 720)
        # Assert that the function returned False after executing the FFmpeg process
        mock_print.assert_called_with("Resizing video to 720p...")
        # Assert that the remove_file function was called
        self.assertFalse(result)

if __name__ == '__main__':
    unittest.main()


tests/test_utils.py
from unittest.mock import patch, MagicMock
from src.ColabKit import utils
import unittest
import os


class UtilsTestCase(unittest.TestCase):

    file_path = "tests/file.txt"
    file_not_exists = "non_existing/file.txt"
    # Create a new file
    with open(f"./{file_path}", 'w') as file:
        pass

    def test_existing_file(self):
        # Test case for an existing file
        result = utils.file_exists(self.file_path)
        self.assertTrue(result)

    def test_non_existing_file(self):
        # Test case for a non-existing file
        result = utils.file_exists(self.file_not_exists)
        self.assertFalse(result)

    def test_directory(self):
        # Test case for a directory
        route = "/tests"
        result = utils.file_exists(route)
        self.assertFalse(result)

    def test_remove_existing_file(self):
        utils.remove_file(self.file_path)
        self.assertFalse(os.path.exists(self.file_path))

    def test_remove_non_existing_file(self):
        # Test removing a non-existing file
        utils.remove_file(self.file_not_exists)
        self.assertFalse(os.path.exists(self.file_not_exists))


    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 100, 100])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_unchanged_file(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=3)
        # Assert that the result is True
        self.assertTrue(result)

    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 200, 200])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_changed_file(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=3)
        # Assert that the result is True
        self.assertTrue(result)

    @patch('os.path.exists', return_value=False)
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_non_existing_file(self, mock_sleep, mock_exists):
        # Test case for a non-existing file
        result = utils.wait_for_file('non_existing_file.txt', time_limit=3)
        # Assert that the result is False
        self.assertFalse(result)

    @patch('os.path.exists', side_effect=[False, True, True, True])
    @patch('os.path.getsize', side_effect=[0, 100, 100, 100])
    @patch('time.sleep', new_callable=MagicMock)
    def test_wait_for_existing_unchanged_file_with_custom_time_limit(self, mock_sleep, mock_getsize, mock_exists):
        # Test case for an existing file
        result = utils.wait_for_file('test_file.txt', time_limit=2)
        # Assert that the result is True
        self.assertTrue(result)

if __name__ == '__main__':
    unittest.main()


tests/test_widget.py
from unittest.mock import MagicMock, patch
from src.ColabKit.widget import generate_button
import ipywidgets as widgets
import unittest

class WidgetTestCase(unittest.TestCase):

    text = "My Button"
    tip = "Click here"
    style = "success"
    func = MagicMock()
    param = (1, 2, 3)

def test_generate_button_creation(self):
    with patch('ipywidgets.Button', new_callable=MagicMock) as mock_button:
        # Create the button and call the function with the given arguments
        generate_button(self.text, self.tip, self.style, self.func, *self.param)
        # Assert that the Button class was called with the expected arguments
        mock_button.assert_called_with(description=self.text, button_style=self.style, tooltip=self.tip, icon="erase")
            
    def test_generate_button_function_execution(self):
        # Create button with specified parameters
        button = generate_button(self.text, self.tip, self.style, self.func, *self.param)
        # Simulate button click
        button.click()
        # Verify that the function was called with the expected parameters
        self.func.assert_called_with(*self.param)
    

if __name__ == '__main__':
    unittest.main()

src/PyColab/utils.py
if 'google.colab' in globals():
  from google.colab import files, drive
import time
import os

def remove_file(route) -> None:
    """
    Removes a file if it exists.
    Args:
        route (str): The route of the file to be removed.
    Returns:
        None
    """
    if os.path.isfile(route):
        os.remove(route)


def file_exists(route) -> bool:
    """
    Check if a file exists at the given route.
    Args:
        route (str): The route to the file.
    Returns:
        bool: True if the file exists, False otherwise.
    """
    return os.path.exists(route)


def mount_drive() -> None:
    """
    Mounts the Google Drive to the Colab environment.
    Returns:
        None
    """
    if not os.path.isdir("/content/drive/MyDrive"):
        drive.mount('/content/drive', force_remount=True)


def upload_file(route) -> None:
    """
    Uploads a file to the specified route.
    Args:
        route (str): The destination route for the uploaded file.
    Returns:
        None
    """
    remove_file(route)
    uploaded = files.upload()
    for filename in uploaded.keys():
        os.rename(filename, route)


def wait_for_file(file_name: str, time_limit: int = 60) -> bool:
    """
    Wait for a file to exist and remain unchanged for a specific time limit.
    Args:
        file_name (str): The name of the file to wait for.
        time_limit (int, optional): The maximum time to wait in seconds. Defaults to 60.
    Returns:
        bool: True if the file exists and remains unchanged within the time limit, False otherwise.
    """
    initial_time = time.time()
    exists = False

    while time.time() - initial_time < time_limit:
        if os.path.exists(file_name):
            previous_size = os.path.getsize(file_name)
            time.sleep(1)
            current_size = os.path.getsize(file_name)
            if current_size == previous_size:
                exists = True
                break
    return exists


src/PyColab/record.py
from .template import AUDIO_RECORD, SAVED_AUDIO_RECORDING
from scipy.io.wavfile import read as wav_read, write as wav_write
from IPython.display import HTML, display
if 'google.colab' in globals():
    from google.colab.output import eval_js
from .utils import wait_for_file
from base64 import b64decode
import ffmpeg
import io


def get_audio() -> tuple:
    """
    Get audio from user input through the browser.    
    Returns:
        audio (ndarray): Numpy array containing the audio data.
        sr (int): Sample rate of the audio data.
    """
    # Display the audio recording widget in the browser
    display(HTML(AUDIO_RECORD))    
    # Get the audio data from the browser
    data = eval_js("data")
    binary = b64decode(data.split(',')[1])    
    # Convert the binary audio data to a WAV file
    process = (
        ffmpeg
        .input('pipe:0')
        .output('pipe:1', format='wav')
        .run_async(
            pipe_stdin=True,
            pipe_stdout=True,
            pipe_stderr=True,
            quiet=True,
            overwrite_output=True
        )
    )
    output, err = process.communicate(input=binary)    
    # Calculate the size of the RIFF chunk
    riff_chunk_size = len(output) - 8
    q = riff_chunk_size
    b = []
    for i in range(4):
        q, r = divmod(q, 256)
        b.append(r)    
    # Replace the RIFF chunk size in the output
    riff = output[:4] + bytes(b) + output[8:]    
    # Read the WAV file and return the audio data and sample rate
    sr, audio = wav_read(io.BytesIO(riff))
    return audio, sr

def record(file_audio = "record.wav") -> None:
    """
    Records audio using the get_audio function and saves it as a WAV file.
    Waits for the file to be saved and displays a message if successful.
    Args:
        file_audio (str): The name of the WAV file to be saved.
    Returns:
        None
    """
    # Get the audio data and the sample rate
    audio, sr = get_audio()
    # Save the audio as a WAV file
    wav_write(file_audio, sr, audio)
    # Wait for the file to be saved
    if wait_for_file(file_audio):
        # Display a message indicating the audio recording was saved
        display(HTML(SAVED_AUDIO_RECORDING))


src/PyColab/widget.py
from IPython.display import display
import ipywidgets as widgets


def generate_button(text, tip, style, func, *param) -> widgets.Button:
    """
    Generates an interactive button in the Jupyter Notebook environment with the specified functionality.
    Args:
        text (str): The text displayed on the button.
        tip (str): The tooltip text displayed when hovering over the button.
        style (str): The style of the button (e.g., 'success' for a green button).
        func (callable): The function to be executed when the button is clicked.
        *param: Optional parameters to be passed to the function when the button is clicked.
    Returns:
        widgets.Button
    Example:
        To create a button that executes the function 'my_function' with two parameters:
        generate_button("My Button", "Click here", "success", my_function, param1, param2)
    """
    button = widgets.Button(description=text, button_style=style, tooltip=tip, icon='erase')
    def on_button_click(b):
        func(*param)
    button.on_click(on_button_click)
    display(button)
    return button

src/ColabKit/template.py
AUDIO_SHOW = """<audio controls src="%s"></audio>"""

VIDEO_SHOW = """<video controls width=500><source src="%s" type="video/mp4"></video>"""

AUDIO_RECORD = """
<script>
    var my_div = document.createElement("DIV");
    var recordButton = document.createElement("BUTTON");

    // Establecer estilos en línea para el div
    my_div.style.display = "flex";
    my_div.style.flexDirection = "column";
    my_div.style.gap = "10px";
    my_div.style.alignItems = "center";
    my_div.style.padding = "10px";

    // Establecer estilos en línea para el botón
    recordButton.style.backgroundColor = "red";
    recordButton.style.color = "white";
    recordButton.style.margin = "0 auto";
    recordButton.style.border = "none";
    recordButton.style.borderRadius = "5px";
    recordButton.style.padding = "5px 10px";
    recordButton.style.cursor = "pointer";

    my_div.appendChild(recordButton);
    document.body.appendChild(my_div);

    var base64data = 0;
    var reader;
    var recorder, gumStream;

    var handleSuccess = function (stream) {
      gumStream = stream;
      var options = {
        mimeType: 'audio/webm;codecs=opus'
      };
      recorder = new MediaRecorder(stream);
      recorder.ondataavailable = function (e) {
        var url = URL.createObjectURL(e.data);
        var preview = document.createElement('audio');
        preview.controls = true;
        preview.src = url;
        my_div.appendChild(preview);

        reader = new FileReader();
        reader.readAsDataURL(e.data);
        reader.onloadend = function () {
          base64data = reader.result;
        }
      };
      recorder.start();
    };

    recordButton.innerText = "Grabando... presione para detener";

    navigator.mediaDevices.getUserMedia({ audio: true }).then(handleSuccess);


    function toggleRecording() {
      if (recorder && recorder.state == "recording") {
        recordButton.innerText = "Guardando la grabación... ¡espera por favor!"
        recordButton.style.backgroundColor = "yellow";
        recordButton.style.color = "black";
        recorder.stop();
        gumStream.getAudioTracks()[0].stop();
      }
    }

    function sleep(ms) {
      return new Promise(resolve => setTimeout(resolve, ms));
    }

    var data = new Promise(resolve => {
      recordButton.onclick = () => {
        toggleRecording()

        sleep(2000).then(() => {
          resolve(base64data.toString())
        });
      }
    });

  </script>
"""

SAVED_AUDIO_RECORDING = """
<script>
recordButton.innerText = "Audio guardado con éxito"
recordButton.style.backgroundColor = "green";
recordButton.style.color = "white";
</script>
"""


src/ColabKit/media.py
from .template import VIDEO_SHOW, AUDIO_SHOW
from .utils import file_exists, remove_file
from IPython.display import display, HTML
from moviepy.editor import VideoFileClip
from base64 import b64encode
import ffmpeg

def showMedia(file_path, media) -> None:
    """
    Display media file on Jupyter Notebook.
    Args:
        file_path (str): The path to the media file.
        media (str): The type of media file ('video' or 'audio').
    Returns:
        None
    """
    # Check if the file exists
    if file_exists(file_path):
        # Set the content type and media type based on the media type
        content_type = 'video/mp4' if media == "video" else 'audio/wav'
        media_type = VIDEO_SHOW if media == "video" else AUDIO_SHOW        
        # Read the file data and encode it as base64
        data = open(file_path, 'rb').read()
        data_url = f"data:{content_type};base64," + b64encode(data).decode()        
        # Display the media file
        display(HTML(media_type % data_url))
    else:
        # Print an error message if the file does not exist
        print(f"\033[43m\033[101mThe {media} file {file_path} is not found in the directory.\033[0m")

def video_duration(video_path) -> float:
    """
    Returns the duration of a video.
    :Args:
        video_path (str) The path to the video file.
    :return: 
        (float) The duration of the video in seconds.
    """
    return VideoFileClip(video_path).duration


def get_video_resolution(video_path) -> tuple:
    """
    Get the resolution of a video file.    
    Args:
        video_path (str): The path to the video file.    
    Returns:
        tuple: A tuple containing the width and height of the video.
    """
    video = VideoFileClip(video_path)
    return (video.size[0], video.size[1])


def ffmpeg_err(stderr) -> str:
    """
    Extracts and formats error lines from the stderr output of the ffmpeg command.
    Args:
        stderr (bytes): The stderr output of the ffmpeg command.
    Returns:
        str: A formatted string containing the error lines.
    """
    # Convert the stderr output to a string and split it into lines
    lines = stderr.decode('utf-8').splitlines()
    # Filter out the lines that start with "Error"
    error_lines = [line for line in lines if line.startswith("Error")]
    # Format the error lines and return them
    for error_line in error_lines:
        return f"\033[43m\033[101m{error_line}\033[0m"


def ffmpeg_proc(output_stream, output_file, msg) -> bool:
    """
    Run FFmpeg with the specified output stream and file.
    Parameters:
    output_stream (str): The output stream to be processed by FFmpeg.
    output_file (str): The file path where the output will be saved.
    msg (str): A message to be displayed in the console.
    Returns:
    bool: True if FFmpeg ran successfully, False otherwise.
    """
    try:
        # Run FFmpeg with the specified output stream and options
        ffmpeg.run(output_stream, capture_stderr=True, overwrite_output=True)        
        # Print the success message
        print(f'File {msg} and saved as {output_file}')        
        # Return True to indicate success
        return True
    except ffmpeg.Error as e:
        # Print the FFmpeg error message
        print(ffmpeg_err(e.stderr))        
        # Remove the output file if an error occurs
        remove_file(output_file)        
        # Return False to indicate failure
        return False

def ffmpeg_conv(input_file: str, output_file: str) -> bool:
    """
    Convert a video file from one format to another using FFmpeg.
    Args:
        input_file: The path to the input video file.
        output_file: The path to save the converted video file.
    Returns:
        bool: True if the conversion was successful, False otherwise.
    """
    # Print the conversion message
    print(f"Converting {input_file} to {output_file}...")
    # Create the input stream
    input_stream = ffmpeg.input(input_file)
    # Create the output stream
    output_stream = ffmpeg.output(input_stream, output_file)
    # Execute the FFmpeg process and return the output
    return ffmpeg_proc(output_stream, output_file, "converted")


def ffmpeg_trim(input_file, start, interval, output_file) -> str:
    """
    Trims a video file using FFmpeg.
    Args:
        input_file (str): The path to the input video file.
        start (int): The starting point in seconds.
        interval (int): The duration in seconds to trim.
        output_file (str): The path to the output trimmed video file.
    Returns:
        str: The path to the trimmed video file.
    """
    # Print the progress
    print(f"Trimming {input_file} to {interval} seconds...")
    # Create the input stream
    input_stream = ffmpeg.input(input_file, ss=start)
    # Create the output stream
    output_stream = ffmpeg.output(input_stream, output_file, t=interval)
    # Run FFmpeg to trim the video
    return ffmpeg_proc(output_stream, output_file, "trimmed")


def resize_video(video_path: str, new_resolution: int):
    """
    Resize a video to a new resolution.
    Args:
        video_path (str): The path to the video file.
        new_resolution (int): The new resolution height in pixels.
    Returns:
        (str, bool): The path to the resized video file if successful, otherwise False.
    """
    # Generate the new path for the resized video
    new_path = f"{video_path.split('.')[0]}_{new_resolution}p.mp4"
    # Get the current resolution of the video
    video_resolution = get_video_resolution(video_path)
    print(f"Video resolution: {video_resolution}")
    if video_resolution[0] >= 1920 or video_resolution[1] >= 1080:
        # Resize the video to the new resolution
        print(f"Resizing video to {new_resolution}p...")
        new_width = 2 * int((video_resolution[0] / video_resolution[1] * new_resolution / 2))
        input_stream = ffmpeg.input(video_path)
        output_stream = ffmpeg.output(
            input_stream, new_path, vf=f'scale={new_width}:{new_resolution}', vcodec='libx264')
        # Check if the resizing was successful
        return new_path if ffmpeg_proc(output_stream, new_path, "resized") else False
    else:
        print("No need to change the size")
        return video_path
